Source/DafnyDriver/LocalTestLoggerEvents.cs

```diff
@@ -0,0 +1,423 @@
+    }
+
+    /// <summary>
+    /// Invokes each of the subscribers of the event and handles exceptions which are thrown
+    /// ensuring that each handler is invoked even if one throws.
+    /// The actual calling of the subscribers is done on a background thread.
+    /// </summary>
+    private void SafeInvokeAsync(Func<MulticastDelegate> eventHandlersFactory, EventArgs args, int size,
+      string traceDisplayName) {
+      ValidateArg.NotNull(eventHandlersFactory, nameof(eventHandlersFactory));
+      ValidateArg.NotNull(args, nameof(args));
+
+      // Invoke the handlers on a background thread.
+      this.loggerEventQueue.QueueJob(
+        () => {
+          var eventHandlers = eventHandlersFactory();
+          eventHandlers?.SafeInvoke(this, args, traceDisplayName);
+        }, size);
+    }
+
+    /// <summary>
+    /// Method called to process a job which is coming from the logger event queue.
+    /// </summary>
+    private void ProcessQueuedJob(Action action) {
+      action();
+    }
+
+    /// <summary>
+    /// Throws if we are disposed.
+    /// </summary>
+    private void CheckDisposed() {
+      if (this.isDisposed) {
+        throw new ObjectDisposedException(typeof(TestLoggerEvents).FullName);
+      }
+    }
+
+    /// <summary>
+    /// The method parses the config file of vstest.console.exe to see if the Max Job Queue Length is defined.
+    /// Return the Max Queue Length so defined or a default value specified by TestPlatformDefaults.DefaultMaxLoggerEventsToCache
+    /// </summary>
+    private int GetMaxNumberOfJobsInQueue() {
+      return GetSetting(TestPlatformDefaults.MaxNumberOfEventsLoggerEventQueueCanHold,
+        TestPlatformDefaults.DefaultMaxNumberOfEventsLoggerEventQueueCanHold);
+    }
+
+    /// <summary>
+    /// The method parses the config file of vstest.console.exe to see if the Max Job Queue size is defined.
+    /// Return the Max Queue size so defined or a default value specified by TestPlatformDefaults.DefaultMaxJobQueueSize
+    /// </summary>
+    private int GetMaxBytesQueueCanHold() {
+      return GetSetting(TestPlatformDefaults.MaxBytesLoggerEventQueueCanHold,
+        TestPlatformDefaults.DefaultMaxBytesLoggerEventQueueCanHold);
+    }
+
+    /// <summary>
+    /// Returns whether flow control on logger events queue should be enabled or not. Default is enabled.
+    /// </summary>
+    private static bool IsBoundsEnabledOnLoggerEventQueue() {
+      bool enableBounds;
+#if NETFRAMEWORK
+            string enableBoundsOnEventQueueIsDefined =
+ ConfigurationManager.AppSettings[TestPlatformDefaults.EnableBoundsOnLoggerEventQueue];
+#else
+      string enableBoundsOnEventQueueIsDefined = null;
+#endif
+      if (string.IsNullOrEmpty(enableBoundsOnEventQueueIsDefined)) {
+        enableBounds = TestPlatformDefaults.DefaultEnableBoundsOnLoggerEventQueue;
+      } else {
+        if (!(bool.TryParse(enableBoundsOnEventQueueIsDefined, out enableBounds))) {
```

Comment:
There's an extraneous pair of parentheses on this line (after the `!`).

Reply:
Ditto